### PR TITLE
enabled selecting between serial and OpenMP distance functions

### PR DESCRIPTION
**Pull request by orbeckst**
- all functions in lib.distances acquired the mode=<mode> keyword;
  - "serial" selects the standard serial C/Cython code from lib._distances;
  - "OpenMP" selects the OpenMP parallelized C/Cython code from
    lib._distances_openmp
- on systems without an OpenMP capable compiler, this will likely just mean
  that the OpenMP versions are actually serial (the OpenMP pragma should
  get ignored) --- need to be confirmed/tested
- tests in test_distances were refactored so that both serial and OpenMP
  versions are tested
- the current implementation is not very pretty (see lib.distances.select())
  but will make it very easy to add additional acceleration schemes such as
  CUDA or numba as long as the call signatures of the calc_* functions remain
  the same.

### CODE REVIEW

**Comment by orbeckst** [comment]
Btw, I am very open for any comments on how to make this nicer, and I am also not very happy with the name *mode* for the keyword argument as used to determine if running in serial or OpenMP, e.g

```python
d = MDAnalysis.lib.distances.distance_array(g.positions, ow.positions, box=u.dimensions, 
                                            mode="OpenMP")
```

**Comment by kain88-de** [comment]
@orbeckst instead of `mode` you could use a boolean switch `parallel=True/False`. Or if the openMP code supports a variable number of jobs/CPUs some packages have adopted a `jobs` argument, stating the number of CPU's used. `-1` is then means use all available CPU's.

**Comment by jbarnoud** [comment]
I agree with @kain88-de : `mode` is a rather vague keyword.

File: package/MDAnalysis/lib/distances.py [hunk 1]
```diff
@@ -24,19 +24,60 @@
...
+_distances['serial'] = importlib.import_module("._distances",
+                                         package="MDAnalysis.lib")
+try:
+    _distances['OpenMP'] = importlib.import_module("._distances_openmp",
+                                          package="MDAnalysis.lib")
+except ImportError:
+    pass
+del importlib
+
+def select(funcname, args=None, kwargs=None, mode="serial"):
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/distances.py]: Could we make this _select so people don't try and use directly?

File: package/MDAnalysis/lib/distances.py [hunk 2]
```diff
@@ -24,19 +24,60 @@
...
 
 from .mdamath import triclinic_vectors, triclinic_box
+
+
+# hack to select backend with mode=<mode> kwarg. Note that
+# the cython parallel code (prange) in paralle.distances is
+# independent from the OpenMP code
+import importlib
+_distances = {}
+_distances['serial'] = importlib.import_module("._distances",
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/distances.py]: With these different backends, is it possible to cache the available functions when they are first imported?  So each "module" is a dict of functions

File: testsuite/MDAnalysisTests/test_distances.py [hunk 3]
```diff
@@ -63,12 +67,20 @@ def mindist(a, b, box):
...
 
-        assert_almost_equal(val, ref, decimal=6, err_msg="Issue 151 not correct (PBC in distance array)")
+        assert_almost_equal(val, ref, decimal=6,
+                            err_msg="Issue 151 not correct (PBC in distance array)")
 
+class TestDistanceArray_Serial(_TestDistanceArray):
+    mode = "serial"
 
-class TestDistanceArrayDCD(TestCase):
+class TestDistanceArray_OpenMP(_TestDistanceArray):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/test_distances.py]: I'm not 100% sure that travis will a) build in openmp and b) run with more than one thread (and find race conditions)

But it's still good to have the tests obviously

**Comment by richardjgowers** [comment]
Maybe backend instead of mode?

**Comment by jbarnoud** [comment]
`backend` looks clearer already.  I would go for `parallel`. 
Le 11 nov. 2015 18:54, Richard Gowers <notifications@github.com> a écrit :Maybe backend instead of mode?

**Comment by orbeckst** [comment]
On 11 Nov, 2015, at 09:54, Richard Gowers wrote:

> Maybe backend instead of mode?

Funnily enough, that's exactly what I just changed it to. It's a bit technical but the best description I can think of. parallel is not good because parallel="serial" sounds really strange.

With backend we can also imagine backend="CUDA", backend="OpenCL", backend="SPIDAL" etc.

**Comment by hainm** [comment]
I am +1 for backend if you have more than 2 choices.

File: package/MDAnalysis/lib/distances.py (hunk 2 above)
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: I think that would be possible; in principle you should be able to use the `module.__dict__` for that (looks a bit hackish --- is this something one would do?) However, keeping it a function allows better control over errors.

In any case, this is a purely internal thing and we can play with it later, too.

File: testsuite/MDAnalysisTests/test_distances.py (hunk 3 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/test_distances.py]: Is there a way for travis to report what code we actually get? Or can we add a function or variable to the modules that reports on what is baked in, e.g.
```python
_distances.BACKEND == "serial"
_distances.BACKEND == "OpenMP"  # if compiled with OpenMP
_distances.BACKEND == "serial"  # if OpenMP was not available but we built it anyway
```

File: package/MDAnalysis/lib/distances.py (hunk 1 above)
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: I changed it to `_run()`.

File: testsuite/MDAnalysisTests/test_distances.py (hunk 3 above)
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/test_distances.py]: That's actually a really good point, currently we have no way of knowing how the Cython was compiled from the module.

Some attribute like `MDAnalysis.lib._distances_openmp.USED_OPENMP` would save us some headaches in the future.

I'll have a look at how we can do this.  Not a showstopper though.

**Comment by orbeckst** [comment]
Rebased to include
* `mode` --> `backend`
* `select()` --> `_run`

Also removed `cimport c_numpy` (see mailing list for [additional discussion](https://groups.google.com/d/msg/mdnalysis-devel/OmKnXvEROGA/8O0VtRbjAwAJ)).

**Comment by hainm** [comment]
I am not sure why duplicate a lot of code here

```bash
$ diff distances_openmp.pyx distances.pyx 
19,20c19,20
< Parallel distance calculation library --- :mod:`MDAnalysis.lib._distances_openmp`
< =================================================================================
---
> Distance calculation library --- :mod:`MDAnalysis.lib._distances`
> =================================================================
22,23c22
< 
< Contains OpenMP versions of the contents of "calc_distances.h"
---
> Serial versions of all distance calculations
25a25,28
> cimport cython
> cimport c_numpy
> c_numpy.import_array()
> 
49a53
>     void minimum_image(double *x, float *box, float *inverse_box)
83c87
<                                    <double*>result.data) 
---
>                                    <double*>result.data)
255a260,307
> 
> 
> @cython.boundscheck(False)
> def contact_matrix_no_pbc(coord, sparse_contacts, cutoff):
>     cdef int rows = len(coord)
>     cdef double cutoff2 = cutoff ** 2
>     cdef float[:, ::1] coord_view = coord
> 
>     cdef int i, j
>     cdef double[3] rr;
>     cdef double dist
>     for i in range(rows):
>         sparse_contacts[i, i] = True
>         for j in range(i+1, rows):
>             rr[0] = coord_view[i, 0] - coord_view[j, 0]
>             rr[1] = coord_view[i, 1] - coord_view[j, 1]
>             rr[2] = coord_view[i, 2] - coord_view[j, 2]
>             dist = rr[0]*rr[0] + rr[1]*rr[1] + rr[2]*rr[2]
>             if dist < cutoff2:
>                 sparse_contacts[i, j] = True
>                 sparse_contacts[j, i] = True
> 
> 
> @cython.boundscheck(False)
> def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
>     cdef int rows = len(coord)
>     cdef double cutoff2 = cutoff ** 2
>     cdef float[:, ::1] coord_view = coord
>     cdef float[::1] box_view = box
>     cdef float[::1] box_inv = 1. / box
> 
>     cdef int i, j
>     cdef double[3] rr;
>     cdef double dist
>     for i in range(rows):
>         sparse_contacts[i, i] = True
>         for j in range(i+1, rows):
>             rr[0] = coord_view[i, 0] - coord_view[j, 0]
>             rr[1] = coord_view[i, 1] - coord_view[j, 1]
>             rr[2] = coord_view[i, 2] - coord_view[j, 2]
> 
>             minimum_image(rr, &box_view[0], &box_inv[0])
> 
>             dist = rr[0]*rr[0] + rr[1]*rr[1] + rr[2]*rr[2]
> 
>             if dist < cutoff2:
>                 sparse_contacts[i, j] = True
>                 sparse_contacts[j, i] = True
```

File: package/MDAnalysis/lib/distances.py [hunk 4]
```diff
@@ -24,19 +24,60 @@
...
+
+# hack to select backend with backend=<backend> kwarg. Note that
+# the cython parallel code (prange) in parallel.distances is
+# independent from the OpenMP code
+import importlib
+_distances = {}
+_distances['serial'] = importlib.import_module("._distances",
+                                         package="MDAnalysis.lib")
+try:
+    _distances['OpenMP'] = importlib.import_module("._distances_openmp",
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/distances.py]: This is pretty minor, but can we make the backend selection case insensitive?  I think you could just call this 'openmp' and then call `.lower()` on whatever string is passed.

**Comment by orbeckst** [comment]
With OpenMP-versions available I can now implement #530 and remove `lib.parallel.distances`.

(This comment mainly serves to link to #530.)

**Pull request merged**